app/services/GuidanceService.scala

```diff
@@ -20,54 +20,73 @@ import connectors.GuidanceConnector
 import javax.inject.{Inject, Singleton}
 import models.ui.{FormData, PageContext}
 import play.api.Logger
+import models.errors._
+import models.RequestOutcome
 import uk.gov.hmrc.http.HeaderCarrier
 import scala.concurrent.{ExecutionContext, Future}
 import repositories.SessionRepository
 import models.ocelot.Process
 
 @Singleton
 class GuidanceService @Inject() (connector: GuidanceConnector, sessionRepository: SessionRepository, pageBuilder: PageBuilder, uiBuilder: UIBuilder) {
-  val logger: Logger = Logger(getClass)
+  val logger = Logger(getClass)
 
-  def getPageContext(url: String, sessionId: String, formData: Option[FormData] = None)(implicit context: ExecutionContext): Future[Option[PageContext]] =
-    sessionRepository.get(sessionId).map { processOption =>
-      processOption.flatMap { process =>
+  def getPageContext(url: String, sessionId: String, formData: Option[FormData] = None)(
+      implicit context: ExecutionContext
+  ): Future[RequestOutcome[PageContext]] =
+    sessionRepository.get(sessionId).map {
+      case Right(process) =>
         pageBuilder
           .pages(process)
           .fold(
             err => {
               logger.warn(s"PageBuilder error $err for url $url on process ${process.meta.id}")
-              None
+              Left(InvalidProcessError)
             },
             pages => {
               val stanzaIdToUrlMap: Map[String, String] = pages.map(page => (page.id, s"/guidance${page.url}")).toMap
               pages
                 .find(page => page.url == url)
-                .map(pge => PageContext(uiBuilder.fromStanzaPage(pge, formData)(stanzaIdToUrlMap), s"/guidance${pages.head.url}"))
+                .fold(Left(NotFoundError): RequestOutcome[PageContext]) { pge =>
```

Comment:
I'm not quite certain what to use here. The url comes from the redirect so I wouldn't expect this to fail. Maybe use internal server error as this should not happen.

Reply:
The Url could have come from someone just trying random urls after a session has started